src/caf/distribute/gravity_model/multi_area.py

```diff
@@ -0,0 +1,695 @@
+        be accounted for, and should only appear in one distribution each.
+    """
+
+    # actual cost matrix
+    # adjusted cost matrix (optional) instead could be callable
+
+    # Need to update functions in base class to allow for giving multiple cost distributions
+    # Ideally calibrate and calibrate_with_perceived factors should be defined flexibly
+    # in base class
+
+    def __init__(
+        self,
+        row_targets: np.ndarray,
+        col_targets: np.ndarray,
+        cost_matrix: np.ndarray,
+        cost_function: cost_functions.CostFunction,
+        params: Optional[MultiDistInput],
+    ):
+        super().__init__(cost_function=cost_function, cost_matrix=cost_matrix)
+        self.row_targets = row_targets
+        self.col_targets = col_targets
+        if len(row_targets) != cost_matrix.shape[0]:
+            raise IndexError("row_targets doesn't match cost_matrix")
+        if len(col_targets) != cost_matrix.shape[1]:
+            raise IndexError("col_targets doesn't match cost_matrix")
+        if params is not None:
+            self.tlds = pd.read_csv(params.TLDFile)
+            self.tlds.rename(
+                columns={
+                    params.cat_col: "cat",
+                    params.min_col: "min",
+                    params.max_col: "max",
+                    params.ave_col: "ave",
+                    params.trips_col: "trips",
+                },
+                inplace=True,
+            )
+            self.lookup = pd.read_csv(params.TldLookupFile)
+            self.lookup.rename(
+                columns={params.lookup_zone_col: "zone", params.lookup_cat_col: "cat"},
+                inplace=True,
+            )
+            self.tlds.set_index("cat", inplace=True)
+            self.lookup.sort_values("zone")
+            self.init_params = params.init_params
+            self.dists = self.process_tlds()
+            self.log_path = params.log_path
+            self.furness_tol = params.furness_tolerance
+            self.furness_jac = params.furness_jac
+
+    def process_tlds(self):
+        """Get distributions in the right format for a multi-area gravity model"""
+        dists = []
+        for cat in self.tlds.index.unique():
+            tld = self.tlds.loc[cat]
+            tld = cost_utils.CostDistribution(tld)
+            zones = self.lookup[self.lookup["cat"] == cat].index.values
+            if len(zones) == 0:
+                raise ValueError(
+                    f"{cat} doesn't seem to appear in the given tld "
+                    "lookup. Check for any typos (e.g. lower/upper case). "
+                    f"If this is expected, remove {cat} from your "
+                    "tlds dataframe before inputting."
+                )
+
+            distribution = MultiCostDistribution(
+                name=cat, cost_distribution=tld, zones=zones, function_params=self.init_params
+            )
+            dists.append(distribution)
+
+        return dists
```

Comment:
If this function is entirely about translating some cost distributions that are in some odd format into one that is suitable for the cost distribution to use, does it not make more sense for the gravity model to take the output format of this function as it's expected input?

We can then add this function to a class that formats and stores the data?